src/can.rs

```diff
@@ -0,0 +1,192 @@
+//! Controller Area Network
+
+/// CAN Identifier
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Id {
+    /// Standard 11bit Identifier (0..=0x7FF)
+    Standard(u32),
+
+    /// Extended 29bit Identifier (0..=0x1FFF_FFFF)
+    Extended(u32),
+}
+
+impl Id {
+    /// Returs true when the identifier is valid, false otherwise.
+    pub fn valid(self) -> bool {
+        match self {
+            Id::Standard(id) if id <= 0x7FF => true,
+            Id::Extended(id) if id <= 0x1FFF_FFFF => true,
+            _ => false,
+        }
+    }
+}
+
+/// A CAN2.0 Frame
+pub trait Frame: Sized {
+    /// Creates a new frame.
+    /// Returns an error when the the identifier is not valid or the data slice is too long.
+    fn new(id: Id, data: &[u8]) -> Result<Self, ()>;
+
+    /// Creates a new remote frame (RTR bit set).
+    /// Returns an error when the the identifier is  or the data length code (DLC) not valid.
+    fn new_remote(id: Id, dlc: usize) -> Result<Self, ()>;
+
+    /// Returns true if this frame is a extended frame.
+    fn is_extended(&self) -> bool;
+
+    /// Returns true if this frame is a standard frame.
+    fn is_standard(&self) -> bool {
+        !self.is_extended()
+    }
+
+    /// Returns true if this frame is a remote frame.
+    fn is_remote_frame(&self) -> bool;
+
+    /// Returns true if this frame is a data frame.
+    fn is_data_frame(&self) -> bool {
+        !self.is_remote_frame()
+    }
+
+    /// Returns the frame identifier.
+    fn id(&self) -> Id;
+
+    /// Returns the data length code (DLC) which is in the range 0..8.
+    ///
+    /// For data frames the DLC value always matches the length of the data.
+    /// Remote frames do not carry any data, yet the DLC can be greater than 0.
+    fn dlc(&self) -> usize;
+
+    /// Returns the frame data (0..8 bytes in length).
+    fn data(&self) -> &[u8];
+}
+
+/// A CAN interface that is able to transmit and receive frames.
+pub trait Can {
```

Comment:
is it useful to have a blocking implementation to go along with this? (see [1](https://github.com/rust-embedded/embedded-hal/blob/master/src/spi.rs) [2](https://github.com/rust-embedded/embedded-hal/blob/master/src/blocking/spi.rs)